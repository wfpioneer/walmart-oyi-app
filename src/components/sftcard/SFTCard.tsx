import React, { ReactElement, ReactNode } from 'react';
import { Text, View } from 'react-native';
import FontAwesome5Icon from 'react-native-vector-icons/FontAwesome5';

import styles from './SFTCard.styles';
import COLOR from '../../themes/Color';
import Button from '../button/Button';

interface SFTCardProps {
  iconName?: string;
  iconProp?: ReactNode;
  title: string;
  topRightBtnTxt?: string;
  topRightBtnAction?: () => void;
  bottomRightBtnTxt?: string[];
  bottomRightBtnAction?: (index: number) => void;
  children?: ReactNode | ReactElement;
}

const renderBottomRightBtns = (textArray: string[], actionFunc?: (index: number) => void) => {
  const arraySize = textArray.length;
  return (
    <View style={styles.bottomRowContainer}>
      {!actionFunc &&
        <Text style={styles.bottomRowText}>
          {textArray[0]}
        </Text>
      }
      {actionFunc &&
      <View style={styles.bottomRowBtnContainer}>
          {textArray.map((value, index) => {
            return (
              <View key={index} style={styles.bottomRowBtn}>
                <Button
                  type={Button.Type.NO_BORDER}
                  title={value}
                  onPress={() => actionFunc(index)}
                  titleColor={COLOR.MAIN_THEME_COLOR}
                  titleFontSize={14}
                  titleFontWeight={'bold'}
                  height={24}
                />
                {index + 1 !== arraySize &&
                  <Text style={styles.bottomRowDivider}>
                    |
                  </Text>
                }
              </View>
            )
          })}
        </View>
      }
    </View>
  )
}

const SFTCard = (props: SFTCardProps) => {
  const { iconName, iconProp, title, topRightBtnTxt, topRightBtnAction, bottomRightBtnTxt, bottomRightBtnAction, children } = props;

  return (
    <View style={styles.mainContainer} >
      <View style={styles.topRowContainer} >
        <View style={styles.iconTitleContainer}>
          {iconName ?
            <FontAwesome5Icon name={iconName} size={15} color={COLOR.GREY_700} style={styles.icon} />
            :
            <View style={styles.icon}>
              {iconProp}
            </View>
          }
          <Text style={styles.title}>{title}</Text>
        </View>
        {topRightBtnTxt && topRightBtnAction &&
          <Button
            type={Button.Type.NO_BORDER}
            title={topRightBtnTxt}
            onPress={topRightBtnAction}
            titleColor={COLOR.MAIN_THEME_COLOR}
            titleFontSize={14}
            titleFontWeight={'bold'}
            height={24}
          />
        }
      </View>
      { children }
      { bottomRightBtnTxt && renderBottomRightBtns(bottomRightBtnTxt, bottomRightBtnAction)}
    </View>
  )
};

export default SFTCard;
