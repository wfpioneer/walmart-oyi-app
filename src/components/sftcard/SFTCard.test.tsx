import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';
import SFTCard from './SFTCard';

jest.mock('react-native-vector-icons/FontAwesome5', () => 'Icon');

describe('SalesMetrics component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Renders the SFTCard with bottomRightBtnAction', () => {
    const mockBottomRightBtn0Action = jest.fn();
    const { toJSON, getByTestId } = render(
      <SFTCard
        title="SFT-Title"
        subTitle="SFT-SubTitle"
        bottomRightBtnTxt={['bottomRightBtnTxt']}
        bottomRightBtnAction={[mockBottomRightBtn0Action]}
      >
        <Text>Test</Text>
      </SFTCard>
    );
    const bottomRightBtn0 = getByTestId('bottomRightBtn-0');
    fireEvent.press(bottomRightBtn0);

    expect(mockBottomRightBtn0Action).toBeCalled();
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the SFTCard with bottomRightBtnText only', () => {
    const { toJSON } = render(
      <SFTCard
        title="SFT-Title"
        subTitle="SFT-SubTitle"
        bottomRightBtnTxt={['bottomRightBtnTxt']}
      >
        <Text>Test</Text>
      </SFTCard>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the SFTCard with topRightBtnAction and no subtitle', () => {
    const mockTopRightBtnAction = jest.fn();
    const { toJSON, getByTestId } = render(
      <SFTCard
        title="SFT-Title"
        topRightBtnTxt="topRightBtnTxt"
        topRightBtnAction={mockTopRightBtnAction}
      >
        <Text>Test</Text>
      </SFTCard>
    );
    const topRightBtn = getByTestId('topRightBtn');
    fireEvent.press(topRightBtn);

    expect(mockTopRightBtnAction).toBeCalled();
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the SFTCard with topRightBtnAction, no subtitle and iconName', () => {
    const mockTopRightBtnAction = jest.fn();
    const { toJSON } = render(
      <SFTCard
        iconName="iconName"
        title="SFT-Title"
        topRightBtnTxt="topRightBtnTxt"
        topRightBtnAction={mockTopRightBtnAction}
      >
        <Text>Test</Text>
      </SFTCard>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders the SFTCard with topRightBtnAction, no subtitle and iconProp', () => {
    const mockTopRightBtnAction = jest.fn();
    const { toJSON } = render(
      <SFTCard
        iconProp="iconProp"
        title="SFT-Title"
        topRightBtnTxt="topRightBtnTxt"
        topRightBtnAction={mockTopRightBtnAction}
      >
        <Text>Test</Text>
      </SFTCard>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
