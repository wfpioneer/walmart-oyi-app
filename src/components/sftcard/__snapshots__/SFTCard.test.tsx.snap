// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesMetrics component Renders the SFTCard with bottomRightBtnAction 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#ffffffff",
      "marginTop": 8,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#f5f5f5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          SFT-Title
        </Text>
        <Text
          style={
            Object {
              "color": "#9e9e9e",
              "fontSize": 10,
              "marginLeft": 0,
            }
          }
        >
          SFT-SubTitle
        </Text>
      </View>
    </View>
  </View>
  <Text>
    Test
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "padding": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "height": 24,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#00000000",
                  "flex": 1,
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
              testID="bottomRightBtn-0"
            >
              <Text
                style={
                  Object {
                    "color": "#0071e9",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "textAlign": undefined,
                  }
                }
              >
                bottomRightBtnTxt
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`SalesMetrics component Renders the SFTCard with bottomRightBtnText only 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#ffffffff",
      "marginTop": 8,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#f5f5f5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          SFT-Title
        </Text>
        <Text
          style={
            Object {
              "color": "#9e9e9e",
              "fontSize": 10,
              "marginLeft": 0,
            }
          }
        >
          SFT-SubTitle
        </Text>
      </View>
    </View>
  </View>
  <Text>
    Test
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "padding": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#757575",
        }
      }
    >
      bottomRightBtnTxt
    </Text>
  </View>
</View>
`;

exports[`SalesMetrics component Renders the SFTCard with topRightBtnAction and no subtitle 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#ffffffff",
      "marginTop": 8,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#f5f5f5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          SFT-Title
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "height": 24,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#00000000",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="topRightBtn"
        >
          <Text
            style={
              Object {
                "color": "#0071e9",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "bold",
                "textAlign": undefined,
              }
            }
          >
            topRightBtnTxt
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Text>
    Test
  </Text>
</View>
`;

exports[`SalesMetrics component Renders the SFTCard with topRightBtnAction, no subtitle and iconName 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#ffffffff",
      "marginTop": 8,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#f5f5f5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <Icon
        color="#616161"
        name="iconName"
        size={15}
        style={
          Object {
            "marginRight": 16,
          }
        }
      />
      <View>
        <Text
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          SFT-Title
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "height": 24,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#00000000",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="topRightBtn"
        >
          <Text
            style={
              Object {
                "color": "#0071e9",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "bold",
                "textAlign": undefined,
              }
            }
          >
            topRightBtnTxt
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Text>
    Test
  </Text>
</View>
`;

exports[`SalesMetrics component Renders the SFTCard with topRightBtnAction, no subtitle and iconProp 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#ffffffff",
      "marginTop": 8,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#f5f5f5",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <View
        style={
          Object {
            "marginRight": 16,
          }
        }
      >
        iconProp
      </View>
      <View>
        <Text
          style={
            Object {
              "marginLeft": 0,
            }
          }
        >
          SFT-Title
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "height": 24,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#00000000",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="topRightBtn"
        >
          <Text
            style={
              Object {
                "color": "#0071e9",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "bold",
                "textAlign": undefined,
              }
            }
          >
            topRightBtnTxt
          </Text>
        </View>
      </View>
    </View>
  </View>
  <Text>
    Test
  </Text>
</View>
`;
